File: ios/AnyAICLIRemoteCore/Interaction.swift
import Foundation

public enum InteractionKind: String, Equatable {
  case askQuestion = "ask_question"
  case exitPlan = "exit_plan"
}

public struct InteractionOption: Equatable {
  public let label: String
  public let description: String

  public init(label: String, description: String = "") {
    self.label = label
    self.description = description
  }
}

public struct InteractionQuestion: Equatable {
  public let question: String
  public let options: [InteractionOption]
  public let multiSelect: Bool

  public init(question: String, options: [InteractionOption] = [], multiSelect: Bool = false) {
    self.question = question
    self.options = options
    self.multiSelect = multiSelect
  }
}

public struct InteractionAnnotation: Equatable {
  public let notes: String

  public init(notes: String) {
    self.notes = notes
  }
}

public struct PendingInteraction: Identifiable, Equatable {
  public let rpcID: Int
  public let kind: InteractionKind
  public let sessionIdentity: SessionIdentity
  public let toolCallID: String
  public let questions: [InteractionQuestion]
  public let planContent: String
  public let mode: String
  public var id: Int { rpcID }

  public init(
    rpcID: Int,
    kind: InteractionKind,
    sessionIdentity: SessionIdentity,
    toolCallID: String,
    questions: [InteractionQuestion] = [],
    planContent: String = "",
    mode: String = ""
  ) {
    self.rpcID = rpcID
    self.kind = kind
    self.sessionIdentity = sessionIdentity
    self.toolCallID = toolCallID
    self.questions = questions
    self.planContent = planContent
    self.mode = mode
  }
}

public enum InteractionAnswer: Equatable {
  case accept(answers: [String: [String]], annotations: [String: InteractionAnnotation])
  case chatAbout(partialAnswers: [String: String])
  case skipInterview(partialAnswers: [String: String])
  case approve
  case cancel(feedback: String)
  case cancelAsk
  case abandon
}

File: android/core/model/src/main/java/com/anyaicliremote/core/model/Interaction.kt
package com.anyaicliremote.core.model

/** Whether a pending interaction is a question or a plan-approval request. */
enum class InteractionKind { ASK_QUESTION, EXIT_PLAN }

/** One selectable choice for an interaction question. */
data class InteractionOption(
    val label: String,
    val description: String = "",
)

/** One question with its choices. */
data class InteractionQuestion(
    val question: String,
    val options: List<InteractionOption> = emptyList(),
    val multiSelect: Boolean = false,
)

/** Free-text operator note attached to one answered question. */
data class InteractionAnnotation(val notes: String)

/**
 * A provider-neutral interaction the agent is blocking on. The client renders it
 * and sends a neutral answer keyed by [rpcId]; it never parses a provider wire.
 * For an ask, [questions] is populated. For an exit-plan, [planContent] is the
 * plan text (may be empty).
 */
data class PendingInteraction(
    val rpcId: Long,
    val kind: InteractionKind,
    val sessionIdentity: SessionIdentity,
    val toolCallId: String,
    val questions: List<InteractionQuestion> = emptyList(),
    val planContent: String = "",
    val mode: String = "",
)

/**
 * The operator's neutral answer to a [PendingInteraction]. It is encoded to the
 * daemon's neutral response shape; the client never builds a provider payload.
 */
sealed interface InteractionAnswer {
    /** Ask answer: question index -> selected option labels, plus optional per-question notes. */
    data class Accept(
        val answers: Map<String, List<String>>,
        val annotations: Map<String, InteractionAnnotation> = emptyMap(),
    ) : InteractionAnswer
    data class ChatAbout(val partialAnswers: Map<String, String>) : InteractionAnswer
    data class SkipInterview(val partialAnswers: Map<String, String>) : InteractionAnswer

    /** Dismiss an ask without answering. */
    data object CancelAsk : InteractionAnswer

    /** Exit-plan answers. */
    data object Approve : InteractionAnswer
    data class Cancel(val feedback: String = "") : InteractionAnswer
    data object Abandon : InteractionAnswer
}
